File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokens/Token.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens;

import java.util.ArrayList;
import com.truveta.opentoken.attributes.AttributeExpression;

/**
 * A token is a collection of attribute expressions that are concatenated
 * together to get the token signature
 * and a unique identifier.
 */
public interface Token {

    /**
     * Represents the value for tokens with invalid or missing data.
     * This constant is a placeholder for cases where a token is either
     * not provided or contains invalid information.
     */
    public static final String BLANK = "0000000000000000000000000000000000000000000000000000000000000000";

    /**
     * Get the unique identifier for the token.
     * 
     * @return the unique identifier for the token
     */
    String getIdentifier();

    /**
     * Get the list of attribute expressions that define the token.
     * 
     * @return the list of attribute expressions that define the token
     */
    ArrayList<AttributeExpression> getDefinition();
}

File: lib/python/opentoken/src/main/opentoken/tokens/token.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from abc import ABC, abstractmethod
from typing import List
from opentoken.attributes.attribute_expression import AttributeExpression


class Token(ABC):
    """
    A token is a collection of attribute expressions that are concatenated
    together to get the token signature and a unique identifier.
    """

    # Represents the value for tokens with invalid or missing data.
    # This constant is a placeholder for cases where a token is either
    # not provided or contains invalid information.
    BLANK = "0000000000000000000000000000000000000000000000000000000000000000"

    @abstractmethod
    def get_identifier(self) -> str:
        """
        Get the unique identifier for the token.

        Returns:
            The unique identifier for the token.
        """
        pass

    @abstractmethod
    def get_definition(self) -> List[AttributeExpression]:
        """
        Get the list of attribute expressions that define the token.

        Returns:
            The list of attribute expressions that define the token.
        """
        pass
